`no-plusplus` with `'allowForLoopAfterthoughts': true` fails if the afterthought contains multiple `++` or `--` statements

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

<!--
    If you are using ESLint v6.5.0 or later, you can run ESLint with the `--env-info` flag and paste the output here.
-->

* **ESLint Version:** 6.8.0
* **Node Version:** 10.15.3
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**
`babel-eslint`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
module.exports = {
  env: {
    browser: true,
    es6: true,
    jest: true,
    node: true,
    mocha: true,
  },
  extends: [
    'airbnb',
    'eslint:recommended',
    'google',
    'plugin:react/recommended',
    'prettier',
    'prettier/react',
    'plugin:prettier/recommended',
    'plugin:flowtype/recommended',
  ],
  parser: 'babel-eslint',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
      modules: true,
    },
  },
  plugins: ['babel', 'react', 'prettier', 'sort-imports-es6-autofix', 'flowtype'],
  rules: {
    // specifically allowing no-plusplus for for loop afterthoughts
    'no-plusplus': ['error', { 'allowForLoopAfterthoughts': true }],

    // ...all the other rules that aren't relevant to this example
  },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
/* eslint no-plusplus: ["error", { "allowForLoopAfterthoughts": true }] */
// (actually specified in the project's .eslintrc.js, but here for visibility)

const example = [1, 2, 3, 4, 5];

// eslint errors at `i++` and `j++` here
for (let i = 0, j = i + 1; j < example.length; i++, j++) {
  // ...do stuff with example[i] and example[j]
}
```

<!-- Paste the command you used to run ESLint: -->
```bash
$ eslint --ext .js src
```

**What did you expect to happen?**
Expected eslint not to error on `i++` and `j++` because `{ 'allowForLoopAfterthoughts': true }` is specified

**What actually happened? Please include the actual, raw output from ESLint.**
```bash
/path/to/example
  7:48  error  Unary operator '++' used  no-plusplus
  7:53  error  Unary operator '++' used  no-plusplus
```

I was able to fix the warning by changing `i++, j++` -> `i += 1, j += 1`, but this is something of an antipattern.

Yes, there are other ways to write this, i.e.
```js
const example = [1, 2, 3, 4, 5];

// eslint does not fail at `i++` here
for (let i = 0; i < example.length - 1; i++) {
  // ...do stuff with example[i] and example[i + 1]
}

// or

// eslint does not fail at `i++` here
for (let i = 1; i < example.length; i++) {
  // ...do stuff with example[i - 1] and example[i]
}
```
work just fine, but:
a) I'm updating eslint on a large codebase, so I'd prefer to make as few changes as possible.
b) refactors for this simple example aside, this feels like a bug to me.

If this is the intended behaviour because as touched on in the [no-plusplus docs](https://eslint.org/docs/rules/no-plusplus) there are unintended consequences, then I'd suggest updating that doc to specifically call out that you can only override this rule if the for loop afterthought consists _only_ of one single `++` or `--`. (for example, changing the for loop to the hideous `...; i += 1, j++)` as an experiment still errors on `j++` because the afterthought is a larger expression)

--- a/docs/rules/no-plusplus.md
+++ b/docs/rules/no-plusplus.md
@@ -71,10 +71,30 @@ Examples of **correct** code for this rule with the `{ "allowForLoopAfterthought
 /*eslint no-plusplus: ["error", { "allowForLoopAfterthoughts": true }]*/
 
 for (i = 0; i < l; i++) {
-    return;
+    doSomething(i);
 }
 
-for (i = 0; i < l; i--) {
-    return;
+for (i = l; i >= 0; i--) {
+    doSomething(i);
 }
+
+for (i = 0, j = l; i < l; i++, j--) {
+    doSomething(i, j);
+}
+```
+
+Examples of **incorrect** code for this rule with the `{ "allowForLoopAfterthoughts": true }` option:
+
+```js
+/*eslint no-plusplus: ["error", { "allowForLoopAfterthoughts": true }]*/
+
+for (i = 0; i < l; j = i++) {
+    doSomething(i, j);
+}
+
+for (i = l; i--;) {
+    doSomething(i);
+}
+
+for (i = 0; i < l;) i++;
 ```

--- a/lib/rules/no-plusplus.js
+++ b/lib/rules/no-plusplus.js
@@ -23,15 +23,22 @@ function isForStatementUpdate(node) {
 
 /**
  * Determines whether the given node is considered to be a for loop "afterthought" by the logic of this rule.
- * In particular, if it is the update node of a `ForStatement`, or an operand of a sequence expression that is the update node.
+ * In particular, it returns `true` if the given node is either:
+ *   - The update node of a `ForStatement`: for (;; i++) {}
+ *   - An operand of a sequence expression that is the update node: for (;; foo(), i++) {}
+ *   - An operand of a sequence expression that is child of another sequence expression, etc.,
+ *     up to the sequence expression that is the update node: for (;; foo(), (bar(), (baz(), i++))) {}
  * @param {ASTNode} node The node to check.
  * @returns {boolean} `true` if the node is a for loop afterthought.
  */
 function isForLoopAfterthought(node) {
     const parent = node.parent;
 
-    return isForStatementUpdate(node) ||
-        parent.type === "SequenceExpression" && isForStatementUpdate(parent);
+    if (parent.type === "SequenceExpression") {
+        return isForLoopAfterthought(parent);
+    }
+
+    return isForStatementUpdate(node);
 }
 
 //------------------------------------------------------------------------------

--- a/lib/rules/no-plusplus.js
+++ b/lib/rules/no-plusplus.js
@@ -7,6 +7,34 @@
 "use strict";
 
 //------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Determines whether the given node is the update node of a `ForStatement`.
+ * @param {ASTNode} node The node to check.
+ * @returns {boolean} `true` if the node is `ForStatement` update.
+ */
+function isForStatementUpdate(node) {
+    const parent = node.parent;
+
+    return parent.type === "ForStatement" && parent.update === node;
+}
+
+/**
+ * Determines whether the given node is considered to be a for loop "afterthought" by the logic of this rule.
+ * In particular, if it is the update node of a `ForStatement`, or an operand of a sequence expression that is the update node.
+ * @param {ASTNode} node The node to check.
+ * @returns {boolean} `true` if the node is a for loop afterthought.
+ */
+function isForLoopAfterthought(node) {
+    const parent = node.parent;
+
+    return isForStatementUpdate(node) ||
+        parent.type === "SequenceExpression" && isForStatementUpdate(parent);
+}
+
+//------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
 
@@ -42,18 +70,19 @@ module.exports = {
     create(context) {
 
         const config = context.options[0];
-        let allowInForAfterthought = false;
+        let allowForLoopAfterthoughts = false;
 
         if (typeof config === "object") {
-            allowInForAfterthought = config.allowForLoopAfterthoughts === true;
+            allowForLoopAfterthoughts = config.allowForLoopAfterthoughts === true;
         }
 
         return {
 
             UpdateExpression(node) {
-                if (allowInForAfterthought && node.parent.type === "ForStatement") {
+                if (allowForLoopAfterthoughts && isForLoopAfterthought(node)) {
                     return;
                 }
+
                 context.report({
                     node,
                     messageId: "unexpectedUnaryOp",

--- a/tests/lib/rules/no-plusplus.js
+++ b/tests/lib/rules/no-plusplus.js
@@ -24,7 +24,11 @@ ruleTester.run("no-plusplus", rule, {
 
         // With "allowForLoopAfterthoughts" allowed
         { code: "var foo = 0; foo=+1;", options: [{ allowForLoopAfterthoughts: true }] },
-        { code: "for (i = 0; i < l; i++) { console.log(i); }", options: [{ allowForLoopAfterthoughts: true }] }
+        { code: "for (i = 0; i < l; i++) { console.log(i); }", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (var i = 0, j = i + 1; j < example.length; i++, j++) {}", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; i--, foo());", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; foo(), --i);", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; foo(), ++i, bar);", options: [{ allowForLoopAfterthoughts: true }] }
     ],
 
     invalid: [
@@ -78,6 +82,61 @@ ruleTester.run("no-plusplus", rule, {
                 messageId: "unexpectedUnaryOp",
                 data: {
                     operator: "++"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (i++;;);",
+            options: [{ allowForLoopAfterthoughts: true }],
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "++"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (;--i;);",
+            options: [{ allowForLoopAfterthoughts: true }],
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "--"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (;;) ++i;",
+            options: [{ allowForLoopAfterthoughts: true }],
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "++"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (;; i = j++);",
+            options: [{ allowForLoopAfterthoughts: true }],
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "++"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (;; i++, f(--j));",
+            options: [{ allowForLoopAfterthoughts: true }],
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "--"
                 },
                 type: "UpdateExpression"
             }]

--- a/tests/lib/rules/no-plusplus.js
+++ b/tests/lib/rules/no-plusplus.js
@@ -28,7 +28,11 @@ ruleTester.run("no-plusplus", rule, {
         { code: "for (var i = 0, j = i + 1; j < example.length; i++, j++) {}", options: [{ allowForLoopAfterthoughts: true }] },
         { code: "for (;; i--, foo());", options: [{ allowForLoopAfterthoughts: true }] },
         { code: "for (;; foo(), --i);", options: [{ allowForLoopAfterthoughts: true }] },
-        { code: "for (;; foo(), ++i, bar);", options: [{ allowForLoopAfterthoughts: true }] }
+        { code: "for (;; foo(), ++i, bar);", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; i++, (++j, k--));", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; foo(), (bar(), i++), baz());", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; (--i, j += 2), bar = j + 1);", options: [{ allowForLoopAfterthoughts: true }] },
+        { code: "for (;; a, (i--, (b, ++j, c)), d);", options: [{ allowForLoopAfterthoughts: true }] }
     ],
 
     invalid: [
@@ -54,6 +58,16 @@ ruleTester.run("no-plusplus", rule, {
         },
         {
             code: "for (i = 0; i < l; i++) { console.log(i); }",
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "++"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (i = 0; i < l; foo, i++) { console.log(i); }",
             errors: [{
                 messageId: "unexpectedUnaryOp",
                 data: {
@@ -137,6 +151,17 @@ ruleTester.run("no-plusplus", rule, {
                 messageId: "unexpectedUnaryOp",
                 data: {
                     operator: "--"
+                },
+                type: "UpdateExpression"
+            }]
+        },
+        {
+            code: "for (;; foo + (i++, bar));",
+            options: [{ allowForLoopAfterthoughts: true }],
+            errors: [{
+                messageId: "unexpectedUnaryOp",
+                data: {
+                    operator: "++"
                 },
                 type: "UpdateExpression"
             }]

